### PR TITLE
Deprecate public services

### LINKED ISSUE #6962: Make public services private
## Feature Request

We have a lot of public services that I think we can make them private, [in Symfony `5.1` there is a tag for deprecating a public service to make it private](https://symfony.com/blog/new-in-symfony-5-1-deprecate-public-services-into-private-services), but since we're supporting Symfony `4.4`, we could maybe [copy the compiler pass as internal class](https://github.com/symfony/symfony/blob/d39596edaba715d4c7b4bbef8efbb1c6c4e916f0/src/Symfony/Component/DependencyInjection/Compiler/AliasDeprecatedPublicServicesPass.php) and add a custom tag to deprecate them.

### PR DESCRIPTION
**Pull request by eerison**
## Subject

I am targeting this branch, because it need to be showed in the current version

Closes #6962

```markdown
### Deprecated
 - Deprecated public services
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Isn't this supposed to be deprecated on 3.x, in order to remove them for the 4.0 release ? cc @franmomu

**Comment by eerison** [comment]
> Isn't this supposed to be deprecated on 3.x, in order to remove them for the 4.0 release ? cc @franmomu

Hey @VincentLanglet 

But sonata 4 will support `Symfony 5` as minimus version?

**Comment by VincentLanglet** [comment]
> > Isn't this supposed to be deprecated on 3.x, in order to remove them for the 4.0 release ? cc @franmomu
> 
> Hey @VincentLanglet
> 
> But sonata 4 will support `Symfony 5` as minimus version?

No, we'll allow symfony 4.4 too.

But if I understand correctly https://github.com/sonata-project/SonataAdminBundle/issues/6962#issue-837210119
The goal is 
- To copy the compiler pass in 3.x
- To deprecate the public services in 3.x
- To remove the public services in 4.0
- To remove the compiler pass in 4.0

Why should we keep the public services for 4.4 ? Dependency injection can be use instead, no ?

**Comment by eerison** [comment]
> > > Isn't this supposed to be deprecated on 3.x, in order to remove them for the 4.0 release ? cc @franmomu
> > 
> > 
> > Hey @VincentLanglet
> > But sonata 4 will support `Symfony 5` as minimus version?
> 
> No, we'll allow symfony 4.4 too.
> 
> But if I understand correctly [#6962 (comment)](https://github.com/sonata-project/SonataAdminBundle/issues/6962#issue-837210119)
> The goal is
> 
> * To copy the compiler pass in 3.x
> * To deprecate the public services in 3.x
> * To remove the public services in 4.0
> * To remove the compiler pass in 4.0
> 
> Why should we keep the public services for 4.4 ? Dependency injection can be use instead, no ?

I see, make sense ...

**Comment by franmomu** [comment]
Yes, the idea I had was to add the compiler pass to be able to deprecate public services in `3.x` and also in `4.x` (until we remove support for Symfony `4.4`).

> But will Sonata 4 support Symfony 5 as minimus version?

We will support Symfony 4.4 until there is another LTS version (and we add support).

**Comment by eerison** [comment]
> Yes, the idea I had was to add the compiler pass to be able to deprecate public services in `3.x` and also in `4.x` (until we remove support for Symfony `4.4`).
> 
> > But will Sonata 4 support Symfony 5 as minimus version?
> 
> We will support Symfony 4.4 until there is another LTS version (and we add support).

But if we will remove all `public services` in `sonata 4` then I guess we can remove this `compiler pass` in master, can't we?

**Comment by VincentLanglet** [comment]
> > Yes, the idea I had was to add the compiler pass to be able to deprecate public services in `3.x` and also in `4.x` (until we remove support for Symfony `4.4`).
> > > But will Sonata 4 support Symfony 5 as minimus version?
> > 
> > 
> > We will support Symfony 4.4 until there is another LTS version (and we add support).
> 
> But if we will remove all `public services` in `sonata 4` then I guess we can remove this `compiler pass` in master, can't we?

To me yes, but reading 
> also in `4.x` (until we remove support for Symfony `4.4`)

I'm not sure. @franmomu Why can't we remove the public services in 4.0 ?

**Comment by franmomu** [comment]
If we can make them all private yes, just that I haven't checked if we can remove them all.

**Comment by eerison** [comment]
> If we can make them all private yes, just that I haven't checked if we can remove them all.

if we'll use symfony 4 as the minimus version, IMO we could remove service as `public`

**Comment by eerison** [comment]
Screenshot 2021-04-05 at 20 12 27

it'll show messages like this.

**Comment by VincentLanglet** [comment]
Can you confirm that you did something special to get the deprecation or are they always displayed (cause the deprecation should only be displayed if the deprecated alias is used) ?

**Comment by eerison** [comment]
> Can you confirm that you did something special to get the deprecation or are they always displayed (cause the deprecation should only be displayed if the deprecated alias is used) ?

Hey @VincentLanglet nope, it's showing to all services as public!

But I can do something to show when user use any `service` direct from container

**Comment by VincentLanglet** [comment]
> But I can do something to show when user use any `service` direct from container

It would be great indeed

**Comment by franmomu** [comment]
> Screenshot 2021-04-05 at 20 12 27
> 
> it'll show messages like this.

These messages are maybe because what the compiler pass does is having a `sonata.admin.helper` service for example, it creates:
- `.sonata.container.private.sonata.admin.helper` service
- `sonata.admin.helper` deprecated alias to `.sonata.container.private.sonata.admin.helper` service.

And we have a `Sonata\AdminBundle\Admin\AdminHelper` deprecated alias pointing to the deprecated `sonata.admin.helper`.

Maybe those messages are because of that. I don't know if there is a way to get the aliases of a service to "redirect" them to the non-deprecated service or we have to manually do something.

By the way, I think actions, as well as controllers, need to be public because they are fetched from the container or at least before was that way.

**Comment by VincentLanglet** [comment]
Hi @eerison, did you find a solution ? :)

**Comment by eerison** [comment]
> Hi @eerison, did you find a solution ? :)

Hey @VincentLanglet 
I didn't have time to work on that yet 😢

**Comment by eerison** [comment]
Hey @VincentLanglet 

I had few problems with xdebug, but it's done 🙏🏼 

ok, I back to test this feature, and the code just pass inside of `src/DependencyInjection/Compiler/AliasDeprecatedPublicServicesCompilerPass.php` when I'm registering the service, and not when I try to get using `$container->get()`;

**Comment by eerison** [comment]
@VincentLanglet what do you think we create some configuration like `sonata_public_services` with default value `true`, and the user can set this to `false` when remove all `$container->get` from the code, after change this, all deprecating messages related with it, will disappear because all services will be `public: false`

make sense?

**Comment by eerison** [comment]
> This could be a solution.
> And we'll add a `NEXT_MAJOR` comment to remove the option and declare all the services as private.

exactly!

ok then I'll do this way 😃

**Comment by eerison** [comment]
Hey @VincentLanglet 

are there any way to get values from `src/DependencyInjection/Configuration.php` inside of  `src/Resources/config/actions.php` or `src/DependencyInjection/Compiler/AliasDeprecatedPublicServicesCompilerPass.php` ?

**Comment by VincentLanglet** [comment]
This way https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php#L57 ?

File: src/SonataAdminBundle.php [hunk 1]
```diff
@@ -61,6 +62,7 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new TwigStringExtensionCompilerPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
         $container->addCompilerPass(new AdminMakerCompilerPass());
         $container->addCompilerPass(new AddAuditReadersCompilerPass());
+        $container->addCompilerPass(new AliasDeprecatedPublicServicesCompilerPass());
```
Comment (franmomu) [inline src/SonataAdminBundle.php:65]:
```suggestion
        $container->addCompilerPass(new AliasDeprecatedPublicServicesCompilerPass(), PassConfig::TYPE_AFTER_REMOVING);
```

I think with this we can probably remove all the previous messages, the one from the screenshot.

**Comment by franmomu** [comment]
I think there is no need for adding a new configuration option, if the compiler pass runs only after removing the unused services, in theory, there should be only deprecations from the public services used in `CRUDController` (I think that's the only one using `$container->get()` of these services) and for those we could create an alias to be used in `CRUDController` like `sonata.admin.pool.do-not-use-it` like Symfony has.

**Comment by eerison** [comment]
Ok, I guess it works almost as expected 😄 

after I apply the @franmomu suggestion 

```php
PassConfig::TYPE_AFTER_REMOVING
```

I'm getting few `service alias is deprecated` messages 
Screenshot 2021-04-20 at 18 34 22

and I added  `$this->container->get('sonata.admin.action.dashboard');` in `listAction`

I'm getting two more message

Screenshot 2021-04-20 at 18 42 41


@VincentLanglet and @franmomu what do you think ?

_just to help you find `service alias is deprecated` messages_

Screenshot 2021-04-20 at 18 46 03

**Comment by franmomu** [comment]
Nice, so I guess these deprecations are because we get them from the container in `CRUDController`, what we can do (for `sonata.admin.pool` as example) is in core.php:

```
->alias('.sonata.admin.pool.do-not-use', 'sonata.admin.pool')
    ->public()
```

and in `CRUDController` change `$this->container->get('sonata.admin.pool')` to `$this->container->get('.sonata.admin.pool.do-not-use')`.

Use this strategy for the public services we need for `CRUDController` (in `4.x` they won't be needed).

**Comment by eerison** [comment]
> Nice, so I guess these deprecations are because we get them from the container in `CRUDController`, what we can do (for `sonata.admin.pool` as example) is in core.php:
> 
> ```
> ->alias('.sonata.admin.pool.do-not-use', 'sonata.admin.pool')
>     ->public()
> ```
> 
> and in `CRUDController` change `$this->container->get('sonata.admin.pool')` to `$this->container->get('.sonata.admin.pool.do-not-use')`.
> 
> Use this strategy for the public services we need for `CRUDController` (in `4.x` they won't be needed).

then do I do this in this Pull request?

**Comment by franmomu** [comment]
Yes, the idea is to not have deprecation messages triggered by us after this PR

**Comment by eerison** [comment]
Hey @franmomu it's not woking, I'm getting this error:

```
The controller for URI "/admin/app/user/list" is not callable: The "sonata.admin.pool.do-not-use" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.
```

has it used in other part of sonata?

**Comment by franmomu** [comment]
> Hey @franmomu it's not woking, I'm getting this error:
> 
> ```
> The controller for URI "/admin/app/user/list" is not callable: The "sonata.admin.pool.do-not-use" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.
> ```
> 
> has it used in other part of sonata?

That's in your project? I'll see if I can test your branch later

**Comment by eerison** [comment]
was missing add `public()` after alias, now it's working 🥳

**Comment by franmomu** [comment]
Can you try to see if you get deprecations going to the dashboard? I think actions should be public.

**Comment by eerison** [comment]
> Can you try to see if you get deprecations going to the dashboard? I think actions should be public.

yes I get in the dashboard, I'm on that now ;)

**Comment by eerison** [comment]
I'm getting in dashboard 

```
User Deprecated: The "sonata.admin.action.dashboard" service alias is deprecated. You should stop using it, as it will be removed in the future.
```

and

```
User Deprecated: The "sonata.admin.block.admin_list" service alias is deprecated. You should stop using it, as it will be removed in the future.
```

do you know where `sonata.admin.action.dashboard` and `sonata.admin.block.admin_list` are defined in Dashboard?

**Comment by eerison** [comment]
I found where `sonata.admin.action.dashboard` were used, is missing only `sonata.admin.block.admin_list`

Screenshot 2021-04-20 at 21 02 47


any idea where is it added ?

**Comment by eerison** [comment]
I guess it's in other bundle

```
sonata_block:
  blocks:
    sonata.admin.block.admin_list:
      contexts: [admin]
```

The ``Admin`` list block
------------------------

The ``Admin`` list is a ``Block`` that fetches information from the ``Admin`` service's
``Pool`` and prints it in the nicely formatted list you have on your default Dashboard.
The ``Admin`` list is defined by the ``sonata.admin.block.admin_list`` service, which is
implemented by the ``Block\AdminListBlockService`` class. It is then rendered using the
``@SonataAdmin/Block/block_admin_list.html.twig`` template file.


could you confirm that?

File: src/Command/CreateClassCacheCommand.php [hunk 2]
```diff
@@ -70,7 +70,7 @@ public function execute(InputInterface $input, OutputInterface $output)
 
         $output->write('<info>Writing cache file ...</info>');
         ClassCollectionLoader::load(
-            include ($classmap),
+            include($classmap),
```
Comment (franmomu) [inline src/Command/CreateClassCacheCommand.php:73]: Please revert this and rebase with `3.x`, this will be fixed.

File: src/DependencyInjection/Compiler/AliasDeprecatedPublicServicesCompilerPass.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+use Sonata\AdminBundle\Util\BCDeprecationParameters;
+use Symfony\Component\DependencyInjection\Compiler\AbstractRecursivePass;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @TODO: Remove it when dropping support for Symfony 4.4 and use "container.private" tag instead.
+ * This file is copied from Symfony since it was added in Symfony 5.1 and we are supporting 4.4
```
Comment (franmomu) [inline src/DependencyInjection/Compiler/AliasDeprecatedPublicServicesCompilerPass.php:24]: If we are going to remove it `4.x`, please replace this with a `NEXT_MAJOR: Remove this file.` comment.

File: src/Resources/config/actions.php [hunk 4]
```diff
@@ -28,7 +29,9 @@
     $containerConfigurator->services()
 
         ->set('sonata.admin.action.dashboard', DashboardAction::class)
+            // NEXT_MAJOR: Remove public and sonata.container.private tag.
             ->public()
+            ->tag(AliasDeprecatedPublicServicesCompilerPass::PRIVATE_TAG_NAME, ['version' => '3.x'])
```
Comment (franmomu) [inline src/Resources/config/actions.php:34]: The deprecation you were receiving when going to the dashboard I think it's because they need to be public because they (controller or actions) are fetched from the container, so please revert the changes from `actions.php`, they are fine as public and they will be in `4.0`.

File: src/Resources/config/block.php [hunk 5]
```diff
@@ -23,7 +24,9 @@
     $containerConfigurator->services()
 
         ->set('sonata.admin.block.admin_list', AdminListBlockService::class)
+            // NEXT_MAJOR: Remove public and sonata.container.private tag.
             ->public()
+            ->tag(AliasDeprecatedPublicServicesCompilerPass::PRIVATE_TAG_NAME, ['version' => '3.x'])
```
Comment (franmomu) [inline src/Resources/config/block.php:29]: Based on the comment you made, looks like block services need to be public, I've created https://github.com/sonata-project/SonataBlockBundle/issues/876 to track that.

In the meanwhile, we need block services to be public, so please remove the tags from `block.php`.

File: src/Resources/config/routing/sonata_admin.xml [hunk 6]
```diff
@@ -4,7 +4,7 @@
         <default key="route">sonata_admin_dashboard</default>
         <default key="permanent">true</default>
     </route>
-    <route id="sonata_admin_dashboard" path="/dashboard" controller="sonata.admin.action.dashboard"/>
+    <route id="sonata_admin_dashboard" path="/dashboard" controller="sonata.admin.action.dashboard.do-not-use"/>
```
Comment (franmomu) [inline src/Resources/config/routing/sonata_admin.xml:7]: Revert this because what was explained in `actions.php`.

**Review by franmomu — COMMENTED** [review]
I leave some comments

**Review by franmomu — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @eerison
